### PR TITLE
Search: add isCorrectedSearch() and fix YoutubeSearchExtractor#getSea…

### PR DESCRIPTION
**Pull request by B0pol**
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [x] I have tested the API against [NewPipe](https://github.com/TeamNewPipe/NewPipe).
- [x] I agree to create a pull request for [NewPipe](https://github.com/TeamNewPipe/NewPipe) as soon as possible to make it compatible with the changed API.

See https://github.com/TeamNewPipe/NewPipe/issues/3419 and https://github.com/TeamNewPipe/NewPipe/issues/3419#issuecomment-612453811
With https://github.com/TeamNewPipe/NewPipe/issues/3419, I figured out two things:

YouTube (& YouTube Music) either corrects your query, or suggest another one.
I fixed the suggestion in YouTube which were actually the corrected query, and added a boolean to know if the query was corrected.

### CODE REVIEW

File: extractor/src/main/java/org/schabi/newpipe/extractor/search/SearchExtractor.java [hunk 1]
```diff
@@ -37,4 +37,15 @@ public SearchQueryHandler getLinkHandler() {
...
         return getLinkHandler().getSearchString();
     }
+
+    /**
+     * When you search on some service, it can give you another and corrected request.
+     * This method says if it's the case.
+     * <p>
+     * Example: on YouTube, if you search for "pewdeipie",
+     * it will give you results for "pewdiepie", then isCorrectedSearch should return true.
+     *
```
Comment (wb9688) [inline extractor/src/main/java/org/schabi/newpipe/extractor/search/SearchExtractor.java:47]: Where is there no closing tag?

File: extractor/src/main/java/org/schabi/newpipe/extractor/search/SearchInfo.java [hunk 2]
```diff
@@ -46,6 +47,11 @@ public static SearchInfo getInfo(SearchExtractor extractor) throws ExtractionExc
         } catch (Exception e) {
             info.addError(e);
         }
+        try {
+            info.isCorrectedSearch = extractor.isCorrectedSearch();
```
Comment (wb9688) [inline extractor/src/main/java/org/schabi/newpipe/extractor/search/SearchInfo.java:51]: Please use a setter

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeMusicSearchExtractor.java [hunk 3]
```diff
@@ -141,6 +141,19 @@ public String getSearchSuggestion() throws ParsingException {
...
 
+    @Override
+    public boolean isCorrectedSearch() {
+        final JsonObject itemSectionRenderer = initialData.getObject("contents").getObject("sectionListRenderer")
+                .getArray("contents").getObject(0).getObject("itemSectionRenderer");
+        if (itemSectionRenderer == null) {
+            return false;
+        }
+
+        JsonObject showingResultsForRenderer = itemSectionRenderer.getArray("contents").getObject(0)
```
Comment (wb9688) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeMusicSearchExtractor.java:152]: `final`

File: extractor/src/test/java/org/schabi/newpipe/extractor/services/youtube/search/YoutubeSearchExtractorTest.java [hunk 4]
```diff
@@ -132,8 +132,30 @@ public static void setUp() throws Exception {
...
+        }
+
+        @Override public SearchExtractor extractor() { return extractor; }
+        @Override public StreamingService expectedService() { return YouTube; }
+        @Override public String expectedName() { return QUERY; }
+        @Override public String expectedId() { return QUERY; }
+        @Override public String expectedUrlContains() { return "youtube.com/results?search_query=" + QUERY; }
+        @Override public String expectedOriginalUrlContains() { return "youtube.com/results?search_query=" + QUERY; }
+        @Override public String expectedSearchString() { return QUERY; }
+        @Nullable @Override public String expectedSearchSuggestion() { return null; }
```
Comment (wb9688) [inline extractor/src/test/java/org/schabi/newpipe/extractor/services/youtube/search/YoutubeSearchExtractorTest.java:156]: Why no expected search suggestion here?

File: extractor/src/main/java/org/schabi/newpipe/extractor/search/SearchExtractor.java (hunk 1 above)
Comment (B0pol) [reply extractor/src/main/java/org/schabi/newpipe/extractor/search/SearchExtractor.java:47]: HTML () doesn't require to close p tag. The next one closes it.
It's the same thing for Javadoc. Look at [Javadoc guidelines, getImage](https://www.oracle.com/technetwork/java/javase/documentation/index-137868.html), they didn't put a closing p tag. It's closed by `@return`, as mentionned in the guidelines: 
> + The first line that begins with an "@" character ends the description. There is only one description block per doc comment; you cannot continue the description following block tags.

Plus if you watch java source code, for example [java.lang.String](http://www.docjar.com/html/api/java/lang/String.java.html), you can see there are lots of paragraph but no closing tag.

Finally, I didn't even add the p tag, intellij did it with autoformatting, and see this [SO about p tag in javadoc](https://stackoverflow.com/a/5263598/12680950).

File: extractor/src/test/java/org/schabi/newpipe/extractor/services/youtube/search/YoutubeSearchExtractorTest.java (hunk 4 above)
Comment (B0pol) [reply extractor/src/test/java/org/schabi/newpipe/extractor/services/youtube/search/YoutubeSearchExtractorTest.java:156]: As you can see in the class name, `CorrectedSearch`, this query `pewdeipie` is automatically corrected by youtube, youtube gives us results for `pewdiepie`, it's not the same as suggestion, and it's the point of the pr.
Maybe here it's better explained https://github.com/TeamNewPipe/NewPipe/issues/3419#issuecomment-612453811

File: extractor/src/main/java/org/schabi/newpipe/extractor/search/SearchExtractor.java (hunk 1 above)
Comment (wb9688) [reply extractor/src/main/java/org/schabi/newpipe/extractor/search/SearchExtractor.java:47]: I didn't know that about Javadoc. Interesting…

File: extractor/src/test/java/org/schabi/newpipe/extractor/services/youtube/search/YoutubeSearchExtractorTest.java (hunk 4 above)
Comment (wb9688) [reply extractor/src/test/java/org/schabi/newpipe/extractor/services/youtube/search/YoutubeSearchExtractorTest.java:156]: It isn't the same indeed, but it should still return what it's correcting it to.

**Comment by B0pol** [comment]
nothing has changed with the last push, I just messed up with my folders

File: extractor/src/main/java/org/schabi/newpipe/extractor/search/SearchInfo.java [hunk 5]
```diff
@@ -70,4 +76,16 @@ public String getSearchString() {
     public String getSearchSuggestion() {
         return searchSuggestion;
     }
+
+    public boolean isCorrectedSearch() {
+        return this.isCorrectedSearch;
+    }
+
+    public void setCorrectedSearch(boolean correctedSearch) {
+        isCorrectedSearch = correctedSearch;
```
Comment (Stypox) [inline extractor/src/main/java/org/schabi/newpipe/extractor/search/SearchInfo.java:85]:
```suggestion
        this.isCorrectedSearch = correctedSearch;
```
Use `this` here too, for consistency's sake

Comment (wb9688) [reply extractor/src/main/java/org/schabi/newpipe/extractor/search/SearchInfo.java]: For even more consistency, change that `correctedSearch` to `isCorrectedSearch`

**Comment by wb9688** [comment]
Could you rebase?

**Comment by B0pol** [comment]
I'm getting some failing tests so I'll check later

**Comment by wb9688** [comment]
@B0pol: You just don't have the `ParsingException` imported in `MediaCCCSearchExtractor`

**Comment by B0pol** [comment]
> @B0pol: You just don't have the `ParsingException` imported in `MediaCCCSearchExtractor`

The main problem was the new nanojson version. Could you review if this part is ok?

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeMusicSearchExtractor.java [hunk 6]
```diff
@@ -133,7 +133,7 @@ public String getUrl() throws ParsingException {
     public String getSearchSuggestion() throws ParsingException {
         final JsonObject itemSectionRenderer = initialData.getObject("contents").getObject("sectionListRenderer")
                 .getArray("contents").getObject(0).getObject("itemSectionRenderer");
-        if (itemSectionRenderer == null) {
+        if (itemSectionRenderer.equals(JsonUtils.EMPTY_OBJECT)) {
```
Comment (wb9688) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeMusicSearchExtractor.java:136]: That works, however you could just do `.isEmpty()` instead, which would be slightly faster, but mainly is more readable imho

**Comment by B0pol** [comment]
The failing test in YoutubeSeachExtractorTest, inner class PageTest, test duplicatedItemsCheck is failing with:
"Upload date not in the past"
It's unrelated to this PR, and it's because there is a premiere https://www.youtube.com/watch?v=NWjVaAvXWEo. I'll open an issue soon

**Comment by wb9688** [comment]
@B0pol: That will be fixed by #309

**Pull request merged**